core/trino-main/src/main/java/io/trino/operator/project/MergePages.java

```diff
@@ -75,12 +75,59 @@ private MergePages() {}
             WorkProcessor<Page> pages,
             AggregatedMemoryContext memoryContext)
     {
-        return pages.transform(new MergePagesTransformation(
-                types,
-                minPageSizeInBytes,
-                minRowCount,
-                maxPageSizeInBytes,
-                memoryContext.newLocalMemoryContext(MergePages.class.getSimpleName())));
+        List<Type> typeList = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        WorkProcessor.Transformation<Page, Page> mergingTransform;
+        if (typeList.isEmpty()) {
+            // position count only mode
+            mergingTransform = new MergePositionCountPagesTransformation(minRowCount);
+        }
+        else {
+            mergingTransform = new MergePagesTransformation(
+                    typeList,
+                    minPageSizeInBytes,
+                    minRowCount,
+                    maxPageSizeInBytes,
+                    memoryContext.newLocalMemoryContext(MergePages.class.getSimpleName()));
+        }
+        return pages.transform(mergingTransform);
+    }
+
+    private static class MergePositionCountPagesTransformation
+            implements WorkProcessor.Transformation<Page, Page>
+    {
+        private final int minRowCount;
+        private int currentRowCount;
+
+        private MergePositionCountPagesTransformation(int minRowCount)
+        {
+            this.minRowCount = minRowCount;
+        }
+
+        @Override
+        public TransformationState<Page> process(Page inputPage)
+        {
+            if (inputPage == null) {
+                if (currentRowCount == 0) {
+                    return finished();
+                }
+                Page result = new Page(currentRowCount);
+                currentRowCount = 0;
+                return ofResult(result, false);
+            }
+
+            if (currentRowCount == 0 && inputPage.getPositionCount() >= minRowCount) {
+                // reuse input page directly as the output
+                return ofResult(inputPage);
```

Comment:
Why only when `currentRowCount == 0`?
if pages carry no data, we can reorder them freely